lib/scholar/neighbors/utils.ex

```diff
@@ -2,11 +2,89 @@ defmodule Scholar.Neighbors.Utils do
   @moduledoc false
   import Nx.Defn
   require Nx
+  import Scholar.Shared
+
+  defn linear_search(data, query, opts) do
+    k = opts[:num_neighbors]
+    metric = opts[:metric]
+    batch_size = opts[:batch_size]
+
+    type = Nx.Type.merge(to_float_type(data), to_float_type(query))
+
+    distance_fn =
+      case metric do
+        {:minkowski, p} ->
+          fn data, query ->
+            query
+            |> Nx.new_axis(1)
+            |> Nx.subtract(data)
+            |> Nx.pow(p)
+            |> Nx.sum(axes: [2])
+          end
+
+        {:cosine} ->
+          &Scholar.Metrics.Distance.pairwise_cosine/2
+      end
+
+    {query_size, dim} = Nx.shape(query)
+    num_batches = div(query_size, batch_size)
+    leftover_size = rem(query_size, batch_size)
+
+    batches =
+      query
+      |> Nx.slice_along_axis(0, num_batches * batch_size, axis: 0)
+      |> Nx.reshape({num_batches, batch_size, dim})
+
+    {neighbor_indices, neighbor_distances, _} =
+      while {
+              neighbor_indices = Nx.broadcast(Nx.u64(0), {num_batches, batch_size, k}),
+              neighbor_distances =
+                Nx.broadcast(Nx.as_type(:nan, type), {num_batches, batch_size, k}),
+              {
+                data,
+                batches,
+                i = 0
+              }
+            },
+            i < num_batches do
+        batch = batches[i]
+        distances = distance_fn.(data, batch)
+
+        indices = Nx.argsort(distances, axis: 1, type: :u64) |> Nx.slice_along_axis(0, k, axis: 1)
+        distances = Nx.take_along_axis(distances, indices, axis: 1)
+
+        neighbor_indices = Nx.put_slice(neighbor_indices, [i, 0, 0], Nx.new_axis(indices, 0))
+
+        neighbor_distances =
+          Nx.put_slice(neighbor_distances, [i, 0, 0], Nx.new_axis(distances, 0))
+
+        {neighbor_indices, neighbor_distances, {data, batches, i + 1}}
+      end
+
+    neighbor_indices = Nx.reshape(neighbor_indices, {num_batches * batch_size, k})
+    neighbor_distances = Nx.reshape(neighbor_distances, {num_batches * batch_size, k})
+
+    {neighbor_indices, neighbor_distances} =
+      if leftover_size > 0 do
+        leftover = Nx.slice_along_axis(query, query_size - leftover_size, leftover_size, axis: 0)
+        distances = distance_fn.(data, leftover)
+        indices = Nx.argsort(distances, axis: 1, type: :u64) |> Nx.slice_along_axis(0, k, axis: 1)
+        distances = Nx.take_along_axis(distances, indices, axis: 1)
```

Comment:
These lines are common for the main part of the search in the while loop, maybe it is worth moving this into a separate function, but it's up to you since it's just 3 lines

Reply:
I'm considering it. The code still needs some polishing.
I am also thinking of abstracting the whole thing as kind of map over batches. It might be useful for Task 1 of #246 as well.